### ISSUE TITLE
wyhash bad seeds protection

### ISSUE THREAD

**Issue by likern** [ziglang/zig#10758]
### Zig Version

latest

### Steps to Reproduce

WyHash `wyhash_final3` version has `2^33` bad seeds as was reported here https://github.com/ManDeJan/zig-wyhash/issues/2.
Zig implementation doesn't have bad seeds protection.
Can we mitigate automatically by checking provided seed or generating good seed (where we can) at compile time?

Original message
> This hash function is like all MUM hashes vulnerable to some bad seeds, which when randomly selected by chance lead to security vulnerabilites, at least a high number of collisions independent of the key.

> Please add a seed_init function and change the initial seed if you hit one of the bad seeds. See latest wyhash.

> These bad seeds for final 3 (latest) are:

```c
static void wyhash_seed_init(uint64_t &seed) { // yes, 2^33 bad values
  if ((seed & 0x14cc886e) || (seed & 0x1bf4ed84))
    seed++;
}
static void wyhash32_seed_init(uint32_t &seed) {
  if ((seed == 0x429dacdd) || (seed == 0xd637dbf3))
    seed++;
}
```
> The CONDOM 2 variant should also work, but haven't finished testing it yet.
Haven't tested for your older final1, you can use the new smhasher BadSeeds test for it. final 3 is much faster on newer HW though. the magic numbers on final1 and 2 are a bit different, I think.

https://github.com/ManDeJan/zig-wyhash/issues/2#issue-848391569

### Expected Behavior

WyHash has `comptime` bad-seed protection.

### Actual Behavior

No checking provided seed on bad seeds.

**Comment by jedisct1** [comment]
(originally from #2797)

Hi!

And thanks for reporting this!

wyhash is not meant to be a cryptographic hash, not even for hash table indices.

Even for simulations, weak seeds can be an issue. But this is a breaking change, which is incompatible with all existing implementations. If the purpose of using wyhash is to get reproducible results, this is not great.

Is there a coordinated effort with other languages and implementations to address this and tweak the state the same way for these weak keys?

If we change the `init(seed: u64)` function to handle weak keys, we can add an `init_compat(seed: u64)` function that retains the previous behavior.

But that's a change that should be made broadly that just in Zig. Did you get any input from @wangyi-fudan about this?

**Comment by wangyi-fudan** [comment]
Do you trust the God that create the world or the devil that destroy the world?
And I feel like being insulted as ruban invades wyhash's personal space (making use of default internal seed).
Think about it...

**Comment by jedisct1** [comment]
@wangyi-fudan So what would be your recommendation? Keep the status quo? Use blind multiplication?

**Comment by 189900** [comment]
@jedisct1 I believe this can be closed due to the merged update to version 4 (#15969) that doesn't suffer from bad seeds.

**Comment by jedisct1** [comment]
Indeed!
